### ISSUE TITLE
CollectionTally (EquivalentTo) should throw for non-transitive comparisons

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2870]
As @CharliePoole and I were discussing, we think this should probably be merged before @ggeurts' work at https://github.com/nunit/nunit/pull/2830. I marked it as high priority with respect to that PR.

The NUnit team made a design decision in the past to ignore tolerance for `EquivalentTo`. It's such a tough problem to solve and there isn't demand for it, so I don't think we should change that decision right now. But here is an example of code that implements tolerance manually via a non-transitive custom comparer:

```cs
Assert.That(
    new[] { 2f, 1f },
    Is.EquivalentTo(new[] { 3f, 2f })
        .Using<float>((a, b) => Math.Abs(a - b) <= 1));
```

It returns an incorrect answer. If you swap the order of the array elements, it'll pass. But rather than passing or failing, I think the test should error with InvalidOperationException in the [CollectionTally constructor](https://github.com/nunit/nunit/blob/ffcfba9bdd85792f6aa32dada77a4b044857b4bf/src/NUnitFramework/framework/Constraints/CollectionTally.cs#L70) (since CollectionTally can be used directly) if the comparer being used is non-transitive.

In addition, maybe we could hard-obsolete the Using overloads which enable non-transitivity: the ones that will cause InvalidOperationException to be thrown no matter what. This would have to be done carefully because the `Using` overloads are defined in the a base class, but the base class (`CollectionItemsEqualConstraint`) works just fine with non-transitivity, so we can't block that.

```cs
Assert.That(
    new[] { 2f, 1f },
    Is.EquivalentTo(new[] { 3f, 2f })
        .Using<float>((a, b) => Math.Abs(a - b) <= 1));
```

The overloads which allow non-transitivity are:

- [`CollectionItemsEqualConstraint.Using(Comparison<T>)`](https://github.com/nunit/nunit/blob/ffcfba9bdd85792f6aa32dada77a4b044857b4bf/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs#L114)
- [`CollectionItemsEqualConstraint.Using(Func<T, T, bool>)`](https://github.com/nunit/nunit/blob/ffcfba9bdd85792f6aa32dada77a4b044857b4bf/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs#L146)

What are your preferences? Anything we should consider?

**Comment by jnm2** [comment]
Actually, perhaps the it should be `ArgumentException` in case the user is using the constructor directly and not via `Assert.That`.

**Comment by CharliePoole** [comment]
Sounds good to me. A few comments...

1. Title seems wrong. Don't we have to ban all Using statements, since we know nothing of their transitivity?

2. What if we refactor so that CollectionItemsEqualConstraint derived from a base without Using?

**Comment by jnm2** [comment]
In other words, an end user can break the interface's documented contract and build a non-transitive implementation of IEqualityComparer, but that's a .NET BCL training issue for the end user, not an NUnit issue. They would also have to implement GetHashCode which should make it obvious to such a user that they are building an inherent contradiction.

**Comment by OmicronPersei** [comment]
>if the comparer being used is non-transitive

How is this something that can be determined programatically?

**Comment by jnm2** [comment]
@OmicronPersei This:

> Or are we just saying that any comparer specified with `.Using` that __can__ allow non-transitivity, should be disallowed?

We should shadow and obsolete these `Using` methods in `CollectionEquivalentConstraint`:

- [`CollectionItemsEqualConstraint.Using(Comparison<T>)`](https://github.com/nunit/nunit/blob/ffcfba9bdd85792f6aa32dada77a4b044857b4bf/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs#L114)
- [`CollectionItemsEqualConstraint.Using(Func<T, T, bool>)`](https://github.com/nunit/nunit/blob/ffcfba9bdd85792f6aa32dada77a4b044857b4bf/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs#L146)

**Comment by OmicronPersei** [comment]
Oh alright, sounds much easier than I initially thought!

**Comment by OmicronPersei** [comment]
I'm a little confused, I add the `new` keyword to the methods to shadow them but then I get compiler warnings saying that they don't shadow anything.

**Comment by OmicronPersei** [comment]
Guys, any thoughts?  I'd like to work on this but don't really know what to do.

**Comment by mikkelbu** [comment]
I have not looked at this before, but my understanding of https://github.com/nunit/nunit/issues/2870#issuecomment-415233316 is to add the following to `CollectionEquivalentConstraint` (probably with better summary and description 😄)
```csharp
        /// <summary>
        /// TODO
        /// </summary>
        [Obsolete("Don't")]
        public new CollectionItemsEqualConstraint Using<T>(Comparison<T> comparison)
        {
            return base.Using<T>(comparison);
        }

        /// <summary>
        /// TODO
        /// </summary>
        [Obsolete("Don't")]
        public new CollectionItemsEqualConstraint Using<T>(Func<T, T, bool> comparer)
        {
            return base.Using<T>(comparer);
        }
```
and then do some consequential corrections in `CollectionEquivalentConstraintTests.cs`.

**Comment by OmicronPersei** [comment]
That's what I was thinking, however VS generates a warning (which is treated as an error by the compiler):

![image](https://user-images.githubusercontent.com/5589035/48452554-361f2200-e775-11e8-80e8-9f12b80d0d45.png)

**Comment by mikkelbu** [comment]
@OmicronPersei As far as I can tell the problem is that you are making the change in `CollectionItemsEqualConstraint` and not in `CollectionEquivalentConstraint`.

**Comment by OmicronPersei** [comment]
Ah, yeah that would help wouldn't it? :) Thanks!

**Comment by Dreamescaper** [comment]
Personally I use Func overload quite a lot. It's much more convenient than to implement interface somewhere.
Do we really need to obsolete that?
Can we fallback to o(n^2) verification instead for those overloads? I'd say that in most cases performance in insignificant for such verification.

**Comment by jnm2** [comment]
@Dreamescaper

> Can we fallback to o(n^2) verification instead for those overloads?

I suppose, so long as we detect non-transitivity. We would need to error (not fail) because it would mean your comparer was inherently meaningless.

> I'd say that in most cases performance in insignificant for such verification.

Performance for the EquivalentTo constraint has been a priority. (https://github.com/nunit/nunit/issues/2598) However, we could mention that the `Using(IEqualityComparer<>)` overload is the one to use when performance is important.

@nunit/framework-team What do you think? I'm leaning towards thinking that we should provide a slow implementation that error if the comparer behaves intransitively. This will take some thought to get right, but I'm worried that if we don't do this, folks like Oleksandr are going to just move the same potentially non-transitive logic to a `IEqualityComparer` implementation and potentially still be getting false positives or false negatives.

**Comment by jnm2** [comment]
Also, I still think merging https://github.com/nunit/nunit/pull/3106 leaves the master branch better off than it is now. What do you think about obsoleting now and reintroducing these methods once they are safe to use?

**Comment by CharliePoole** [comment]
I've been musing over this issue once again, both as it impacts NUnit and thinking about how I would implement equivalence in a new framework I'm working on. So, let me throw in a bit of "philosophical" design discussion around the question of what we should enable, what we should guarantee and specifially whether we should enable the user to do things that don't work.

For many years, the underlying NUnit philosophy was pretty much "We try to do what the user tells us, whether it makes sense or not. We report an error if it doesn't work." For example, if you are running the console in a 64-bit process and ask NUnit to run a test assembly that needs 32 bits in process, it goes ahead and errors. It doesn't stop and say "That won't work" or "I don't think that will work, would you like to try it anyway?"

There's a discussion to be had about whether this is still appropriate. Maybe the audience for NUnit has changed and a different approach is needed. If so, it will be hard because __if we substitute the program's judgement or the users's, we really need to get it right__.

That's just background. I'm don't want to have that discussion hear and now. But it colors how I think about the `Using` methods and equivalence of collections in general.

Using methods allow the user to specify a definition of equality (or comparison) in various ways, following an interface or a delegate definition. Neither we more .NET places restrictions on what those methods must do. In the case of .NET, there is guidance in the form of documentation regarding the semantics of equality (for example) but the user may ignore them. Specifically, equality is transitive but nothing forces the user to implement it that way. Is allowing them to implement it "wrong" an error? A feature? A guarantee that whatever they do will work?

Consider the possibilities:
  * An implementation that allows for tolerance.
  * An implementation that treats the number 42 as a wild card, equal to all other numbers.
  * An implementation that returns true if both numbers are prime.
  * An implementation that always returns true.
  * An implementation that always returns false.
  * An implementation that throws an exception if either number is equal to 42.

If the user supplies such implementations to scalar test of equality, we simply use it. We make no check on the implementation because (a) it would break things and (b) it would be impossible to guarantee that our check had caught everything anyway.

If the user supplies such an implementation to a collection equality comparison, then we again use it. The same reasons apply plus the result is completely deterministic.

OTOH, for an equivalence comparison, our own implementation requires some things from the user implementation in order to work correctly. @jnm2 has correctly stated that we require transitivity. We also require stability and some reasonable approximation of the meaning of equality for the types in use. So how do we tell users that?

Some options include
 1. Don't allow Using at all when testing for equivalence.
 2. State the requirements in the documentation and warn that the results may be incorrect if those requirements are not met.
 3. Issue an actual warning whenever Using appears in an equivalence test.
 4. Make tolerance testing a specific supported customization of equivalence and create a new algorithm that deals with it correctly. Such an algorithm would have to deal with the possibility that some data sets may not allow for a solution, much as one does in a linear programming algorithm. This is not simple and the potential benefit of each refinement should probably be weighed. We might consider doing this __only__ for floating point values, since that's really the common example that would appear to have the must usage.

Bottom line, I don't have an answer, just a feeling we need to decide what the ultimate goal is before putting in a first step in some direction we may not want to go.

**Comment by mr-russ** [comment]
I've read Charlie's last comment.  I have implemented (4) in the merged pull request.  @OsirisTerje, It's been 4 years without comment on equivalent to.  My usage of the system is I need to understand what I think equal means for the context I'm using it.  I also need to understand what I mean when I as the NUnit extender allow a tolerance.

This is either,
1. Closed as possible to do with tolerances now.
2. Open because it's an ongoing policy discussion.

Given the ages, and lack of additional input in 4 years, I'm inclined to go with (1).

**Comment by OsirisTerje** [comment]
The we do that @mr-russ .
